coderd/workspaceagentapi.go

```diff
@@ -0,0 +1,815 @@
+		}
+
+		var newHealth database.WorkspaceAppHealth
+		switch update.Health {
+		case agentproto.AppHealth_DISABLED:
+			newHealth = database.WorkspaceAppHealthDisabled
+		case agentproto.AppHealth_INITIALIZING:
+			newHealth = database.WorkspaceAppHealthInitializing
+		case agentproto.AppHealth_HEALTHY:
+			newHealth = database.WorkspaceAppHealthHealthy
+		case agentproto.AppHealth_UNHEALTHY:
+			newHealth = database.WorkspaceAppHealthUnhealthy
+		default:
+			return nil, xerrors.Errorf("unknown health status %q for app %q (%q)", update.Health, updateID, old.Slug)
+		}
+
+		// Don't bother updating if the value hasn't changed.
+		if old.Health == newHealth {
+			continue
+		}
+		old.Health = newHealth
+
+		newApps = append(newApps, *old)
+	}
+
+	for _, app := range newApps {
+		err = a.database.UpdateWorkspaceAppHealthByID(ctx, database.UpdateWorkspaceAppHealthByIDParams{
+			ID:     app.ID,
+			Health: app.Health,
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("update workspace app health for app %q (%q): %w", err, app.ID, app.Slug)
+		}
+	}
+
+	err = a.publishWorkspaceUpdate(ctx, &workspaceAgent)
+	if err != nil {
+		return nil, xerrors.Errorf("publish workspace update: %w", err)
+	}
+	return &agentproto.BatchUpdateAppHealthResponse{}, nil
+}
+
+func (a *AgentAPI) UpdateStartup(ctx context.Context, req *agentproto.UpdateStartupRequest) (*agentproto.Startup, error) {
+	workspaceAgent, err := a.agent(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	a.log.Debug(
+		ctx,
+		"post workspace agent version",
+		slog.F("agent_id", workspaceAgent.ID),
+		slog.F("workspace_id", a.workspaceID),
+		slog.F("agent_version", req.Startup.Version),
+	)
+
+	if !semver.IsValid(req.Startup.Version) {
+		return nil, xerrors.Errorf("invalid agent semver version %q", req.Startup.Version)
+	}
+
+	// Validate subsystems.
+	dbSubsystems := make([]database.WorkspaceAgentSubsystem, 0, len(req.Startup.Subsystems))
+	seenSubsystems := make(map[database.WorkspaceAgentSubsystem]struct{}, len(req.Startup.Subsystems))
+	for _, s := range req.Startup.Subsystems {
+		var dbSubsystem database.WorkspaceAgentSubsystem
+		switch s {
+		case agentproto.Startup_ENVBOX:
+			dbSubsystem = database.WorkspaceAgentSubsystemEnvbox
+		case agentproto.Startup_ENVBUILDER:
+			dbSubsystem = database.WorkspaceAgentSubsystemEnvbuilder
+		case agentproto.Startup_EXECTRACE:
+			dbSubsystem = database.WorkspaceAgentSubsystemExectrace
+		default:
+			return nil, xerrors.Errorf("invalid agent subsystem %q", s)
+		}
+
+		if _, ok := seenSubsystems[dbSubsystem]; !ok {
+			seenSubsystems[dbSubsystem] = struct{}{}
+			dbSubsystems = append(dbSubsystems, dbSubsystem)
+		}
+	}
+
+	err = a.database.UpdateWorkspaceAgentStartupByID(ctx, database.UpdateWorkspaceAgentStartupByIDParams{
+		ID:                workspaceAgent.ID,
+		Version:           req.Startup.Version,
+		ExpandedDirectory: req.Startup.ExpandedDirectory,
+		Subsystems:        dbSubsystems,
+		APIVersion:        AgentAPIVersionREST,
```

Comment:
I think this can be added in a future PR

Reply:
Yeah, that's fine to do it in another PR

Reply:
This causes import cycling so I'm not going to move it as it's a temporary thing until the old API is deprecated/removed. Otherwise this needs to go in agentsdk which makes even less sense